org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/codeactions/cvc_complex_type_2_4_aCodeAction.java

```diff
@@ -0,0 +1,106 @@
+import java.util.List;
+import java.util.TreeSet;
+
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4xml.commons.CodeActionFactory;
+import org.eclipse.lsp4xml.dom.DOMDocument;
+import org.eclipse.lsp4xml.dom.DOMElement;
+import org.eclipse.lsp4xml.dom.DOMNode;
+import org.eclipse.lsp4xml.extensions.contentmodel.model.CMElementDeclaration;
+import org.eclipse.lsp4xml.extensions.contentmodel.model.ContentModelManager;
+import org.eclipse.lsp4xml.services.extensions.ICodeActionParticipant;
+import org.eclipse.lsp4xml.services.extensions.IComponentProvider;
+import org.eclipse.lsp4xml.settings.XMLFormattingOptions;
+import org.eclipse.lsp4xml.utils.LevenshteinDistance;
+import org.eclipse.lsp4xml.utils.XMLPositionUtility;
+
+/**
+ * cvc_complex_type_2_4_a
+ */
+public class cvc_complex_type_2_4_aCodeAction implements ICodeActionParticipant {
+
+	private static final float MAX_DISTANCE_DIFF_RATIO = 0.4f;
+
+	@Override
+	public void doCodeAction(Diagnostic diagnostic, Range range, DOMDocument document, List<CodeAction> codeActions,
+			XMLFormattingOptions formattingSettings, IComponentProvider componentProvider) {
+		try {
+			int offset = document.offsetAt(diagnostic.getRange().getStart());
+			DOMNode node = document.findNodeAt(offset);
+			if (node != null && node.isElement()) {
+				// Get element from the diagnostic
+				DOMElement element = (DOMElement) node;
+				DOMElement parentElement = element.getParentElement();
+				String tagName = element.getSuffix();
+				boolean suffixOnly = element.getPrefix() != null;
+
+				// Get the XSD element declaration of the parent element.
+				ContentModelManager contentModelManager = componentProvider.getComponent(ContentModelManager.class);
+				CMElementDeclaration cmElement = contentModelManager.findCMElement(parentElement);
+				if (cmElement != null) {
+					// Collect all possible elements from the parent element upon the offset start
+					// of the element
+					Collection<CMElementDeclaration> possibleElements = cmElement.getPossibleElements(parentElement,
+							element.getStart());
+					
+					// When added to these collections, the names will be ordered alphabetically
+					Collection<String> otherElementNames = new TreeSet<String>(Collator.getInstance());
+					Collection<String> similarElementNames = new TreeSet<String>(Collator.getInstance());
+					
+					// Try to collect similar names coming from tag name
+					for (CMElementDeclaration possibleElement : possibleElements) {
+						String possibleElementName = possibleElement.getName();
+						if (isSimilar(possibleElementName, tagName)) {
+							similarElementNames.add(possibleElementName);
+						}
+						else {
+							otherElementNames.add(possibleElementName);
+						}
+					}
+
+					// Create ranges for the replace.
+					List<Range> ranges = new ArrayList<>();
+					ranges.add(XMLPositionUtility.selectStartTagName(element, suffixOnly));
```

Comment:
I think we should have 2 methods utility:

 * selectStartTagName(element)
 * selectStartLocalName(element)

and not having a boolean (suffixOnly) because it's more easy to understand that you wish to select a local name or a tag name just by reading code (and not understand what suffixOnly means). More with your IDE you can find references on selectStartLocalName to find all code which wants to select a local name.